### PR TITLE
Makka / 1333 - upgrade wallet connect to v2

### LINKED ISSUE #1333: Upgrade WalletConnect web3-provider to ethereum-provider
> https://docs.walletconnect.com/2.0/advanced/migration-from-v1.x/dapps/#web3-provider

> If you are using @walletconnect/web3-provider, we stopped supporting this provider in favor of the improved version published under @walletconnect/ethereum-provider and you will be able to check the latest version on NPM [here](https://npmjs.com/package/@walletconnect/ethereum-provider).

### PR DESCRIPTION
**Pull request by 0xMakka**
## Description

Upgrade wallet connect to v2. 

**Note:** We need to create a projectId on https://cloud.walletconnect.com. I've temporarily created one and added it to enable testing but will remove once the official projectId is set up and added as an env variable.

Tests to cover:
- Login with wallets listed in the login modal
- Logout with all wallets and ensure user is logged out correctly.
- Switch accounts when logged in and ensure wallet switch works as expected.
- Login/Logout with a wallet in the wallet connect list and ensure it connects successfully with the correct address

## Related Ticket

Resolves #1333 

## Checklist

- [X] I have run `npm run build-all` without errors
- [X] I have formatted JS and TS files with `npm run format-all`

### CODE REVIEW

**Comment by 0xtapi** [comment]
Hey there,

I encountered no technical issues using WalletConnect to connect to Ledger Live on both KlimaDAO and Carbonmark. Seems to work as intended. I also tried reconnecting/disconnecting a few times and switching to MetaMask in between to check if this would somehow screw it up - but it didn't. Tested it on Firefox as well just to make sure. No issues.

I did run into a minor problem when I manually refused the connection, however. See screenshot:

cmblue

1. It uses "Carbonmark Blue" instead of "Klima Green" for the button element (which is also the case in the current live environment).
2. Black text on gray background isn't impossible to read, but it's not optimal. Switching to white would be better.

With a few minor design tweaks, this should be good to go!

File: lib/.env.example [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Create a file called .env.local and provide these keys.
+
+# Required for Wallet Connect v2
+NEXT_PUBLIC_WALLETCONNECT_PROJECT_ID=12345
```
Comment (Atmosfearful) [inline lib/.env.example:4]: unfortunately everything in `lib` should remain environment-agnostic so all variables must be passed in :-/

File: lib/constants/index.ts [hunk 2]
```diff
@@ -302,3 +302,7 @@ export const verra = {
 export const goldStandard = {
   projectDetailPage: `${REGISTRIES.GoldStandard.url}/projects/details`,
 };
+
+export const WALLETCONNECT_PROJECT_ID =
+  process.env.NEXT_PUBLIC_WALLETCONNECT_PROJECT_ID ??
```
Comment (Atmosfearful) [inline lib/constants/index.ts:307]: I'll try to get this set up in our environments by EOD today

File: lib/utils/useProvider/index.tsx [hunk 3]
```diff
@@ -199,15 +205,16 @@ export const useProvider = (): Web3ModalState => {
...
     web3state.provider.provider.on("chainChanged", handleChainChanged);
     /** For WalletConnect and Coinbase disconnections */
     web3state.provider.provider.on("disconnect", handleDisconnect);
 
     return () => {
-      web3state.provider.provider.removeListener(
-        "accountsChanged",
-        handleAccountsChanged
+      web3state.provider.provider.removeListener("accountsChanged", (accts) =>
+        handleAccountsChanged(accts)
```
Comment (Atmosfearful) [inline lib/utils/useProvider/index.tsx:217]: T inline arrow function creates a new instance, so the `removeListener` might not be able to find and remove the handler that was instantiated earlier.

That's why I reuse the same reference to `handleAccountsChanged` directly

**Review by Atmosfearful — CHANGES_REQUESTED** [review]

File: lib/.env.example (hunk 1 above)
Comment (0xMakka) [reply lib/.env.example:4]: Ah okay, apologies my bad I should have realised when there was no example env file in the lib directory. I'll remove this file and leave just leave the projectId key in `lib/constants/index.ts`.

File: lib/utils/useProvider/index.tsx (hunk 3 above)
Comment (0xMakka) [reply lib/utils/useProvider/index.tsx:217]: Yeah that makes sense. Typescript was giving me trouble here hence the change to it - will take another look.

**Review by Atmosfearful — APPROVED** [review]

**Comment by Atmosfearful** [comment]
Works great with Torus! And the modal appears on all 3 applications. Nice work!

**Pull request merged**